src/Workspaces/Core/Portable/Storage/SQLite/v2/SQLitePersistentStorage.Accessor.cs

```diff
@@ -68,35 +68,96 @@ public Accessor(SQLitePersistentStorage storage)
+                var hashData = ReadChecksumColumn(dataId, cancellationToken);
                 return hashData != null && checksum == hashData.Value;
             }
 
             [PerformanceSensitive("https://github.com/dotnet/roslyn/issues/36114", AllowCaptures = false)]
-            public Task<Stream?> ReadStreamAsync(TKey key, Checksum? checksum, CancellationToken cancellationToken)
-                => Storage.PerformReadAsync(
-                    static t => t.self.ReadStream(t.key, t.checksum, t.cancellationToken),
-                    (self: this, key, checksum, cancellationToken), cancellationToken);
+            public async Task<Stream?> ReadStreamAsync(TKey key, Checksum? checksum, CancellationToken cancellationToken)
+            {
+                var (succeeded, dataId) = await TryGetDatabaseIdAsync(key, cancellationToken).ConfigureAwait(false);
+                if (!succeeded)
+                    return null;
+
+                return await Storage.PerformReadAsync(
+                    static t => t.self.ReadStream(t.dataId, t.checksum, t.cancellationToken),
+                    (self: this, dataId, checksum, cancellationToken), cancellationToken).ConfigureAwait(false);
+            }
+
+            private async Task<(bool succeeded, TDatabaseId dataId)> TryGetDatabaseIdAsync(TKey key, CancellationToken cancellationToken)
+            {
+                if (Storage._shutdownTokenSource.IsCancellationRequested)
+                    return default;
+
+                // Getting the database ID may end up needing to write to the DB (to create a unique DB key corresponding
+                // to the data-key the client is passing in).  As such, we have to try to acquire the DB-key in a fashion
+                // that has exclusive write access to the DB to prevent locking errors.
+
+                // First, just try to optimistically read the data id from the db.  This can be done with
+                // just a read connection, so we can run concurrently with all other readers.
+                var (succeeded, dataId) = await Storage.PerformReadAsync(
+                    static t => TryGetDatabaseId(t.self, t.key, allowWrite: false, t.cancellationToken),
+                    (self: this, key, cancellationToken), cancellationToken).ConfigureAwait(false);
+
+                if (succeeded)
+                    return (succeeded, dataId);
+
+                // Data id was not in the db.  Switch over to an exclusive write lock to add it.  This will
+                // always happen the first time we write anything new into the DB.  But from that point on
+                // will very rarely be hit as most read requests will be to a key already written in the past.
+                return await Storage.PerformWriteAsync(
+                    static t => TryGetDatabaseId(t.self, t.key, allowWrite: true, t.cancellationToken),
+                    (self: this, key, cancellationToken), cancellationToken).ConfigureAwait(false);
```

Comment:
however, for the totaally cold case (or the case of a new project/doc) we do have to create the db-id association in the db .  so we upgrade to a write lock to do that, and hte fallback to reading after that point so we can continue to be concurrent with other readers.